### PR TITLE
Feature/scalar func args session config

### LINKED ISSUE #13519: Add `SessionConfig` reference to `ScalarFunctionArgs`
### Is your feature request related to a problem or challenge?

- Now that https://github.com/apache/datafusion/pull/13290 is merged (thanks @joseph-isaacs) we have a place to add new 

@Omega359  [noted](https://github.com/apache/datafusion/pull/13290#discussion_r1852255032) that by adding `SessionConfig` to the `ScalarFunctionArgs`   unblock several tasks such as
- https://github.com/apache/datafusion/issues/13212
- https://github.com/apache/datafusion/issues/12892
- https://github.com/apache/datafusion/issues/10744

### Describe the solution you'd like

Add `&SessionConfig` to `ScalarFunctionArgs`, and ideally add a test that shows the config gets through.

### Describe alternatives you've considered

_No response_

### Additional context

We should try and get this done before DataFusion 44 is released so it isn't a breaking change
- #13334

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #13519

## Rationale for this change

Allow udf's to access df config 

## What changes are included in this PR?

Code.

## Are these changes tested?

Existing tests.

## Are there any user-facing changes?

Not specifically, this is covered with the udf signature change in https://github.com/apache/datafusion/pull/13290

### CODE REVIEW

**Comment by Omega359** [comment]
There is a lot of file changes here but most of the important changes are in scalar_function.rs, There is a todo in [expr_simplifier.rs](https://github.com/apache/datafusion/pull/13527/files#diff-01d563682732bd1ec72187bdb36fe196a58e83feddae28059086793856c3497b) that I would like feedback on.

**Comment by alamb** [comment]
I plan to review this carefully tomorrow

**Comment by Omega359** [comment]
> This PR seems to require `config_options` to be cloned many times now. I wonder if it is possible to avoid that 🤔. I took a brief look and it seems to be somewhat challenging as SessionState allows mutable access to the underlying SessionConfig.

Yes, it's a bit annoying. I was tempted to see if I could switch to &'a ConfigOptions everywhere. There is at least one 'real' (vs Arc::clone) clone for every query, possibly more as I haven't checked. 

> Maybe we could change the semantics so that `SessionConfig` has a `Arc<ConfigOptions>` which was cloned when it was modified (`Arc::unwrap_or_clone()` style) 🤔

Certainly possible, I can attempt that. 

> I also think the const evaluator does need the actual correct ConfigOptions for correctness

I was afraid of that. I was avoiding it because of the signature changes it would required just about everywhere which would cause even more headaches for those systems trying to upgrade.

**Comment by alamb** [comment]
Yeah, it is a tricky one for sure

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Comment by Omega359** [comment]
> > Maybe we could change the semantics so that `SessionConfig` has a `Arc<ConfigOptions>` which was cloned when it was modified (`Arc::unwrap_or_clone()` style) 🤔
> 
> Certainly possible, I can attempt that.

@alamb I did a quick attempt at implementing that however it breaks a commonly used method - SessionConfig.options_mut(). Not having that available breaks a bunch of stuff and while switching to SessionConfig.set(..) is quite possible it's not as clean.

Trying with &ConfigOptions in ScalarFunctionExpr leads to lifetime hell in areas I have no idea how to overcome right now.

As much as I want this feature I'm going to put it aside for now

**Comment by Omega359** [comment]
I think this may be ready for review again. For this round I refactored the code to use &ConfigOptions everywhere except for ScalarFunctionExpr so the cost for cloning ConfigOptions is only incurred when creating a scalar UDF.

**Comment by Omega359** [comment]
Examples check failure is transient I believe.

**Comment by alamb** [comment]
> Examples check failure is transient I believe.

I restarted the checks

File: datafusion-examples/examples/composed_extension_codec.rs [hunk 1]
```diff
@@ -71,8 +71,10 @@ async fn main() {
 
     // deserialize proto back to execution plan
     let runtime = ctx.runtime_env();
+    let state = ctx.state();
+    let config_options = state.config_options();
     let result_exec_plan: Arc<dyn ExecutionPlan> = proto
-        .try_into_physical_plan(&ctx, runtime.deref(), &composed_codec)
+        .try_into_physical_plan(&ctx, config_options, runtime.deref(), &composed_codec)
```
Comment (alamb) [inline datafusion-examples/examples/composed_extension_codec.rs:77]: This is an API change, but I think it is required to thread the config options through as each argument is specific

We could potentially improve the `try_into_physical_plan` API (as a follow on PR) to make it easier to update the API in the future using a trait or something like

https://github.com/apache/datafusion/blob/e99e02b9b9093ceb0c13a2dd32a2a89beba47930/datafusion/expr/src/expr_schema.rs#L39-L38




So this would look something like 
```rust
pub trait ProtobufContext { 
  /// return a function registry
  fn function_registry(&self) -> &dyn FunctionRegistry;
  /// return the runtime env
  fn runtime_env(&self) -> &RuntimeEnv;
  /// return the config options
  fn config_options(&self) -> &ConfigOptions;
  /// return extension codec
  fn extension_codec(&self) -> &dyn PhysicalExtensionCodec;
}
```

```rust
impl AsExecutionPlan for protobuf::PhysicalPlanNode {
...
    fn try_into_physical_plan(
        &self,
        registry: &dyn FunctionRegistry,
        config_options: &ConfigOptions,
        runtime: &RuntimeEnv,
        extension_codec: &dyn PhysicalExtensionCodec,
    ) -> Result<Arc<dyn ExecutionPlan>> {

File: datafusion-examples/examples/pruning.rs [hunk 2]
```diff
@@ -187,7 +188,9 @@ impl PruningStatistics for MyCatalog {
 fn create_pruning_predicate(expr: Expr, schema: &SchemaRef) -> PruningPredicate {
     let df_schema = DFSchema::try_from(schema.as_ref().clone()).unwrap();
     let props = ExecutionProps::new();
-    let physical_expr = create_physical_expr(&expr, &df_schema, &props).unwrap();
+    let config_options = ConfigOptions::default();
```
Comment (alamb) [inline datafusion-examples/examples/pruning.rs:191]: Something we could potentially do to make this API slightly easier to use might be to create a static default `ConfigOptions` 

Something like
```rust
impl ConfigOptions {
  /// returns a reference to default ConfigOptions
  pub fn default_singleton() -> &'static ConfigOptions 
}
```

This would then make it easier to return `&ConfigOptions` in various places when only the default was needed

For example, then in `LocalCsvTableFunc` you could avoid having to thread the `ConfigOptions` through as in that example having the actual config options isn't important

File: datafusion/core/src/datasource/listing/helpers.rs [hunk 3]
```diff
@@ -241,6 +241,7 @@ async fn prune_partitions(
     partitions: Vec<Partition>,
     filters: &[Expr],
     partition_cols: &[(String, DataType)],
+    ctx: &SessionState,
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/helpers.rs:244]: I recommend passing `&ConfigOptions` directly here as the function only needs the config, not the entire session state

File: datafusion/physical-expr/src/scalar_function.rs [hunk 4]
```diff
@@ -243,6 +292,7 @@ pub fn create_physical_expr(
             Arc::new(fun.clone()),
             input_phy_exprs.to_vec(),
             return_type,
+            Arc::new(config_options.clone()),
```
Comment (alamb) [inline datafusion/physical-expr/src/scalar_function.rs:295]: This line worries me -- it means each distinct scalar function in the plan will get an entirely new copy of the ConfigOptions.

I think the `Arc` should be passed in as the argument like this (I realize this will be a significant code change) so that the config options are copied at most once per plan

```rust
/// Create a physical expression for the UDF.
pub fn create_physical_expr(
    fun: &ScalarUDF,
    input_phy_exprs: &[Arc<dyn PhysicalExpr>],
    input_schema: &Schema,
    args: &[Expr],
    input_dfschema: &DFSchema,
    config_options: &Arc<ConfigOptions>, // <--- I think this should be an `Arc`
) -> Result<Arc<dyn PhysicalExpr>> {
```

I think that would also make it clearer that physical planning makes a read only copy of the configuration (`Arc<ConfigOptions>`) that is then unchanged during execution

**Review by alamb — APPROVED** [review]
Thank you @Omega359  -- I think this PR is a major step forward and we could merge it as is.

However, I feel strongly there are two things that should be improved soon (if not this PR then follow on ones). I left specific comments about each
1. Don't copy `ConfigOptions` in every call to `ScalarFunction::create_physical_expr` 
2. Add convenience methods to get default `&ConfigOptions` and `Arc<ConfigOptions>` which I think will help people upgrading to the next version of DataFusion quickly migrate their code

File: datafusion/physical-expr/src/scalar_function.rs (hunk 4 above)
Comment (Omega359) [reply datafusion/physical-expr/src/scalar_function.rs:295]: That was my main concern as well. I am unsure that this change should merged in as is to be honest as any fix will be just as disruptive as this PR is api wise.

Pushing the clone higher up the stack is possible but I did run into major issues trying to push it all the way up including all kinds of disruptive changes like changing OptimizerConfig from &ConfigOptions to the Arc version. Doing the opposite - pushing &ConfigOptions all the way down ran into issues with ScalarFunctionExpr and DynEq/DynHash.

Speaking for myself I prefer the latter as I feel it's less disruptive overall ... I just couldn't get it to work before. Changing the signature of DynHash and DynEq traits may allow it to work but I haven't tried that yet. Maybe with some more thought in January I can get it to work but I expect my January time to be pretty limited.

Comment (alamb) [reply datafusion/physical-expr/src/scalar_function.rs:295]: > Pushing the clone higher up the stack is possible but I did run into major issues trying to push it all the way up including all kinds of disruptive changes like changing OptimizerConfig from &ConfigOptions to the Arc version. Doing the opposite - pushing &ConfigOptions all the way down ran into issues with ScalarFunctionExpr and DynEq/DynHash.

If you have an `Arc<ConfigOptions>` you can always get a `&ConfigOptions` by calling `options.as_ref()`

So in other words, I don't know if you need to push the Arc through everywhere (at least at first) -- we could just change the create physical expr. I would love to help with this -- perhaps I can find time later in the week (though I have a few other things going on too)

Another option to avoid cloning `ConfigOptions` for each instance of `ScalarFunction` might be  to add some sort of API / way for the function to communicate "I need the config options"

**Comment by alamb** [comment]
Converting to draft as this PR is accumulating conflicts and it seems like it is not ready for merge yet

File: datafusion/core/Cargo.toml [hunk 5]
```diff
@@ -144,7 +144,7 @@ zstd = { version = "0.13", optional = true, default-features = false }
 
 [dev-dependencies]
 async-trait = { workspace = true }
-criterion = { workspace = true, features = ["async_tokio"] }
+criterion = { workspace = true, features = ["async_tokio", "async_futures"] }
```
Comment (Omega359) [inline datafusion/core/Cargo.toml:147]: This was added so the spm benchmark would run. It's obvious that benchmarks are not being run regularly at all which should be looked into.

Comment (alamb) [reply datafusion/core/Cargo.toml:147]: I hit the same problem (and made the same fix) in 
- https://github.com/apache/datafusion/pull/15203

**Comment by alamb** [comment]
@Omega359  -- I started reviewing this PR today. I am still getting my head around what the API change is / what this would mean for downstream users. I'll try and spend more time over the next week giving it a more in depth look

**Review by alamb — COMMENTED** [review]
Thank you again for this PR @Omega359 

TLDR is I think the feature (config options to functions) is super valuable

My biggest concern is the impact it would have on downstream projects -- I don't have a good handle on how much churn it would create

What I suggest as a next step is:
1. Send a note to the dev list (and maybe drop a link in discord and slack) asking people with such projects to offer opinions
2. Maybe try to upgrade one of the downstream projects (e.g. comet or delta.rs to see what the impact is)

I'll also kick off some benchmarks on PR too

**Comment by alamb** [comment]
I keep thinking about this PR in the back of my head. The idea of being able to customize functions based on configuration options makes total sense to me, but having to plumb it down all the way through to execution is so disruptive on the APIs and adds non trivial overhead to the whole system for just a few functions

One alternate idea I had was to store whatever setting from the ConfigOptions in the function itself. Then prior to execution we could update all the functions that had option specific state. Something like:

```rust
// update config options...
// Update all registered functions:
for name in ctx.udf_names() {
  let old_func = ctx.get_udf(&name);
  if let Some(new_func) = new_func.with_config_options(options) {
    ctx.register_udf(new_func)
  }
}
```

This would make updating config options potentially slower (as now functions would also need to be updated) but I think the change would be much more localized


Another would be to use the existing [`simplify`](https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarUDF.html#method.simplify) function and update [`SimplifyInfo` ](https://docs.rs/datafusion/latest/datafusion/logical_expr/simplify/trait.SimplifyInfo.html) with the config options (likely much less disruptive)

**Comment by Omega359** [comment]
Not sure I'd call a clone or two overhead but I agree with you on the plumbing. It's a lot.

In JVM land I would likely have used a thread local variable to hold a session id and have a static factory to be able to lookup the config based on that session id. While that could technically work here I think @tustvold has advised in the past the thread locals with tokio are not the way to go.

I had thought of the simplify option but if you look at how that is implemented you would still have to thread the options through the Optimizer and would also require a signature change for OptimizerRule, the same as is required for this PR. It would be less invasive overall though.

Your first option though, hmm. It would work well for config options but long term it wouldn't allow session variables. That could be a future problem to solve I guess. I'll look into what it would take to impl this option.

**Comment by alamb** [comment]
Another idea: 

Since [`ExecutionProps`](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.ExecutionProps.html) is already threaded all the way through and is intended to

> Holds per-query execution properties and data (such as statement starting timestamps).

What if you added the necessary state (like "SQL mode") and then copied that through (possibly all the way to ScalarFunctionArgs) 🤔 

https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarFunctionArgs.html

It would still be some plumbing but it might be much less invasive as ExecutionProps is already sent many places -- ExecutionProps is also already very small 🤔

**Comment by Omega359** [comment]
> Another idea:
> 
> Since [`ExecutionProps`](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.ExecutionProps.html) is already threaded all the way through and is intended to
> 
> > Holds per-query execution properties and data (such as statement starting timestamps).
> 
> What if you added the necessary state (like "SQL mode") and then copied that through (possibly all the way to ScalarFunctionArgs) 🤔
> 
> https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.ScalarFunctionArgs.html
> 
> It would still be some plumbing but it might be much less invasive as ExecutionProps is already sent many places -- ExecutionProps is also already very small 🤔

Hmm. 🤔 

It would mean that we would have to copy things at execution time from ConfigOptions to an instance of ExecutionProps (timezone, sql mode, whatever we deem useful) and perhaps others ([Extensions](https://github.com/apache/datafusion/blob/d2fe308eb15b43bf0a9b5a40652252cd93dd227d/datafusion/common/src/config.rs#L762) comes to mind) though I don't think that would be horribly difficult. Perhaps we could use Vec<[ConfigEntry](https://github.com/apache/datafusion/blob/main/datafusion/common/src/config.rs#L728)> for this - or map<key, ConfigEntry> so consumers don't have to iterate the options to find the one they are interested in.

I agree that this approach would be very much less invasive. I actually like this approach a bit more than [your other idea of adding yet another function](https://github.com/apache/datafusion/pull/13527#issuecomment-2789191368) to the UDF's to set the config_options on them. It would also allow for per-query changes since session variables would then be accessible to UDF's.

Unless someone can poke a pretty big hole in this approach I'll start another branch and work on this.

**Comment by Omega359** [comment]
I've spent some time looking at using ExecutionProps for this and while I think it'll work it's still a lot of churn. That churn is largely because of two reasons:

1. We would need to switch from &ExecutionProps to Arc<ExecutionProps> to eliminate cloning (otherwise there is a lot of lifetime issues with ScalarFunctionExpr).
2. OptimizerConfig - this one is strange - it copies some things from ExecutionProps instead of just using it directly. Either this weirdness needs to be expanded or OptimizerConfig needs to have a function to retrieve ExecutionProps ala how it does for ConfigOptions.

Essentially, I think it'll be better than the ConfigOptions approach but not by a huge amount.

**Comment by alamb** [comment]
> OptimizerConfig - this one is strange - it copies some things from ExecutionProps instead of just using it directly. Either this weirdness needs to be expanded or OptimizerConfig needs to have a function to retrieve ExecutionProps ala how it does for ConfigOptions.

I think OptimizerConfig might predate ExecutionProps perhaps 🤔  -- just using ExecutionProps makes sense to me

**Comment by alamb** [comment]
BTW how many fields from ConfigOptions really need to be copied? Can we just add the ones you need for spark? Or do we need a huge pile of them?

**Comment by alamb** [comment]
Interestingly I noticed that @goldmedal  added `ConfigOptions` to the `async UDF` API in https://github.com/apache/datafusion/pull/14837

Maybe we can do something similar with normal scalar functions

**Comment by Omega359** [comment]
I noticed that in the API but I haven't yet had a chance to see how it's implemented. It would be awesome if it's compatible/workable with non-async functions too

**Comment by Omega359** [comment]
The async udf approach using an ExecutionPlan is interesting. I think it's an approach that could work for scalar udfs as well but it would take some time to impl

**Comment by Omega359** [comment]
Closing as #16970 is preferred approach
